src/modules/HepnosDataStore_service.cc

```diff
@@ -1,15 +1,84 @@
-#include "art/Framework/Services/Registry/ServiceDefinitionMacros.h"
 #include "HepnosDataStore.h"
+#include "art/Framework/Services/Registry/ServiceDefinitionMacros.h"
+#include <chrono>
+#include <functional>
 
 namespace icaruswf {
-  HepnosDataStore::HepnosDataStore(Parameters const & ps)
-    : dataStore_{hepnos::DataStore::connect(ps().protocol(), ps().connection_file())}
-  {}
 
-  hepnos::DataStore& HepnosDataStore::getStore() {
+  HepnosDataStore::HepnosDataStore(Parameters const& ps)
+  {
+
+    hepnos_exec_thread_ = std::thread([this] {
+      while (this->active == true) {
+        if (this->work_to_do) {
+          std::cout << "[HEPnOS thread status] working \n";
+          work();
+          this->work_to_do = 0;
+        } else {
+          // std::cout << "[HEPnOS thread status] going back to sleep \n";
+          std::this_thread::sleep_for(std::chrono::milliseconds(10));
+        }
+      }
+    });
+
+    hepnos_exec_thread_.detach();
+
+    std::function<void(void)> f = [&]() {
+      dataStore_ =
+        hepnos::DataStore::connect(ps().protocol(), ps().connection_file());
+      return;
+    };
+    this->set_work_function(f);
+    this->set_work_state();
+    this->wait();
+  }
+
+  /* Let the worker thread execute work function */
+  void
+  HepnosDataStore::wait()
+  {
+    while (true) {
```

Comment:
This busy loop is OK for the first implementation. In the next version, it needs to be replaced with a condition variable so that we're not wasting a core for each art process on this loop (as Jim had said).